File: go/tire-pressure-monitoring-system/sensor.go
package tpms

import (
	"fmt"
	"math/rand"
)

type sensor struct {
	offset         func() int
	samplePressure func() int
}

func (s sensor) popNextPressurePsiValue() int {
	return s.offset() + s.samplePressure()
}

func NewSensor() Sensor {
	return &sensor{
		offset: func() int {
			return 16
		},
		samplePressure: func() int {
			s := rand.NewSource(42)
			r := rand.New(s)
			pressureTelemetryValue := 6 * r.Intn(1) * r.Intn(1)
			fmt.Println(pressureTelemetryValue)
			return pressureTelemetryValue
		},
	}
}

File: TypeScript/tire-pressure-monitoring-system/sensor.ts
// The reading of the pressure value from the sensor is simulated in this implementation.
// Because the focus of the exercise is on the other class.

export default class Sensor {

	public popNextPressurePsiValue() {
		const pressureTelemetryValue = this.samplePressure();

		return this.offset() + pressureTelemetryValue;
	}

	private samplePressure() {
		// placeholder implementation that simulate a real sensor in a real tire
		const pressureTelemetryValue = Math.floor(6 * Math.random() * Math.random());
		return pressureTelemetryValue;
	}

	private offset() { return 16; }

}
